### PR TITLE
Add url type in the url module

### LINKED ISSUE #366: Add function to extract username and password from credential embedded URL in url module
**Describe the problem/challenge you have**
[A description of the current challenge that you are experiencing.]
When handling a credential embedded URL in the form of `https://username:password@URL`, we want to parse the credential information like `username` and `password`, store them into a k8s secret. Also we want to get a URL that doesn't contain the credential so we can store it elsewhere.

**Describe the solution you'd like**
[A clear and concise description of what you want to happen. If applicable a visual representation of the UX (ex: new CLI argument name, or resulting yaml).]

Involve following methods in `url` module:
1. `extract_username`: takes in a URL string, returns the embedded username as a string. If there's no credential, return `""`.
2. `extract_password`: takes in a URL string, returns the embedded password as a string. If there's no credential, return `""`.
3. `trim_credentials`: takes in a URL string, returns the URL that removes the credentials.

**Anything else you would like to add:**
[Additional information that will assist in solving the issue.]

_(see also: thread in [#carvel](https://kubernetes.slack.com/archives/CH8KCCKA5/p1619026581416200))_

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by hanlins**
Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

Introduce type `url` in `url` module, which has three methods:

1. user.name: to extract username in URL.
2. user.password: to extract password in URL.
3. user: return the colon separated username+password tuple. (if password is not present, then just return username).
4. without_user(): generate a URL object that removes user credential information.
5. string(): returns the string representation of the URL object.

Closes #366.

### CODE REVIEW

File: pkg/yttlibrary/url.go
Comment (pivotaljohn) [inline pkg/yttlibrary/url.go]: Where there are domain-specific canonical terms for things, it can be efficient to reuse those terms rather than coin our own.

I see the corresponding spec — [RFC 2396 § 3.2.2](https://www.rfc-editor.org/rfc/rfc2396.html#section-3.2.2) — the name for this combination is: "userinfo". The Go implementation exposes a field called `User`.

What might motivate us to do something different?

Suggestion: `.without_user()`

File: pkg/yttlibrary/url.go
Comment (pivotaljohn) [inline pkg/yttlibrary/url.go]: This value feels to me more like a `starlarkstruct.Struct` than a `starlarkstruct.Module`: `urlValue` _is_ a data structure.

I do see that `libraryValue` has been implemented as a `module`...
https://github.com/vmware-tanzu/carvel-ytt/blob/bc8782e60d1089a4781060d36811fea5bbbb608b/pkg/workspace/library_module.go#L143-L153

Yet, while it's all methods, it too is a value, not a result of an evaluated Starlark Module.

That `TODO` on line 143 is legit: this isn't a pattern we want to propagate.

...

There's a "constructor" for `Structs`:
https://github.com/vmware-tanzu/carvel-ytt/blob/bc8782e60d1089a4781060d36811fea5bbbb608b/vendor/github.com/k14s/starlark-go/starlarkstruct/struct.go#L50-L52

And `Builtin`s (i.e. a function implemented in Go) can be bound to any Starlark value:
https://github.com/vmware-tanzu/carvel-ytt/blob/bc8782e60d1089a4781060d36811fea5bbbb608b/vendor/github.com/k14s/starlark-go/starlark/value.go#L663-L676

File: pkg/yttlibrary/url.go
Comment (pivotaljohn) [inline pkg/yttlibrary/url.go]: Appreciate the defensive copy! 👍🏼  Values are more reliable when immutable.

File: pkg/yttlibrary/url.go
Comment (pivotaljohn) [inline pkg/yttlibrary/url.go]: What do you think about returning a `urlValue`, here, instead?

It keeps open the possibility of a fluent interface.
It would likely be useful to generally provide a `String()` for `urlValue`.

**Review by pivotaljohn — CHANGES_REQUESTED** [review]
This is a significant leveling-up from #367. Thank you for putting in the extra effort to work out how to provide a value.

Unfortunately, the prior art around that work within the `ytt` code is slightly off from what I believe we want here: a value rather than a code module (see comments, below). So, there's a little adjustment to the actual Starlark data structures we'd use to present these `url` values.

File: pkg/template/core/starlark_value.go
Comment (pivotaljohn) [inline pkg/template/core/starlark_value.go:130]: Yes, this is legitimately dead code. Let's keep this deletion. 👍🏼 

- `starlarkstruct.Struct` was used in the very beginning.
- but then the maintainers found it lacking in important ways — mostly because it's not iterable, but others as well... something that `ytt` really needs.
- it was an oversight that this `case` was not pruned.

File: pkg/yttlibrary/url.go
Comment (pivotaljohn) [inline pkg/yttlibrary/url.go]:
```suggestion
      // sufficient "deep copy" because we set the only pointer in the structure to `nil`
	urlVar := *uv.URL
```

File: pkg/yamltemplate/filetests/ytt-library/url.yml
Comment (pivotaljohn) [inline pkg/yamltemplate/filetests/ytt-library/url.yml]:
```suggestion
url_type:
  #@ testURL1 = url.parse("http://admin:password@url.domain:8080")
  #@ testURL2 = testURL1.without_user()
  #@ testURL3 = url.parse("http://admin@url.domain:8080")
  test1: #@ testURL1.string()
  test2: #@ testURL1.user
  test3: #@ testURL1.user.name
  test4: #@ testURL1.user.password
  test5: #@ testURL2.string()
  test6: #@ testURL2.user
  test7: #@ testURL3.string()
  test8: #@ testURL3.user
  test9: #@ testURL3.user.name
  test10: #@ testURL3.user.password

```

As noted in the summary comment, let's follow your initial idea of _not_ inferring a string encoding from a "naked" instance.

This means that there's a cryptic error message complaining "`...unknown type *starlark.Builtin for conversion to go value...`" 
... it's _this_ situation that makes me think we'd get to where we want to go faster and with the needed control if we "rolled our own" out the gate.

**Review by pivotaljohn — COMMENTED** [review]
Seems like a design conversation is in order... but here are some notes...

## The Tale of Two Structs

So, today I learned:
- `ytt` has not been using `starlarkstruct.Struct`... despite that spot of dead code (the very same you expunged in this PR)
- in fact, we've rolled our own: `core.StarlarkStruct` ... for reasons.
- given the set of specific needs, I'm starting to think that starting out with a custom, yet congruent `struct` impl. might be the best move, here...?

and, it would be a bad idea to have two implementations of `struct` floating around in the runtime... it's just asking for strange behavior.

Apologies for inadvertently misleading you to the defunct implementation. 🙏🏼 


## To String or Not To String

`urlValue` is actually the first time we're hatching a "higher-level" type within `ytt`'s flavor of Starlark. All the other types are either primitives or generic data structures.

I got a little over zealous in my desire for a "great UX" when I suggested that a naked `urlValue` instance should imply the stringified form.
- part of the problem here is that for other similar higher-level types, this may not be the right behavior.
- ... and in other (albiet rarer) use-cases for `urlValue`, the structure form might be what's desired, anyway.

So, better to defer making that inference to a time when we can make a better informed decision (i.e. after some use and ideally a couple more higher-level types). 

Concretely, yes, let's include a `.string()` as you had originally designed.

_(details, below)_

File: pkg/yamltemplate/filetests/ytt-library/url.yml
Comment (hanlins) [reply pkg/yamltemplate/filetests/ytt-library/url.yml]: I think we want to make the following changes after having a discussion last Friday:
1. Instead of using the legacy `starlarkstruct.Struct`, we want to converge to `StarlarkStruct`.
2. Instead of returning a confusing error message `...unknown type *starlark.Builtin for conversion to go value...` when referencing `testURL` directly, we want to see things like `testURL cannot be directly referenced, please use testURL.string()`, something more instructive like that.
3. As @jtigger mentioned, involve a new method `string()` to return the string representation of the URL. We do this   1). to accommodate the potential use case that we might need to return an encoded string that contains Unicode characters, 2). we might introduce more module-level types later on, and we might not always want to depend on the struct constructor. We need to have some flexibility for those types of representations.

I'll try and see if it's feasible to address all these requirements in a single changeset.

File: pkg/yamltemplate/filetests/ytt-library/url.yml
Comment (pivotaljohn) [inline pkg/yamltemplate/filetests/ytt-library/url.yml]: (nit) the test failure message is getting kinda hard to see what went wrong; would you mind pulling this into a separate test file, please? (e.g. `url-value.yml`)?

File: pkg/yttlibrary/url.go
Comment (pivotaljohn) [inline pkg/yttlibrary/url.go]: (nit)

```suggestion
	return nil, fmt.Errorf("urlValue: cannot coerce to a string; use .string()")
```

Attempting to follow the error message formatting I'm seeing elsewhere in Starlark/ytt.

**Review by pivotaljohn — COMMENTED** [review]
This is fantastic, Hanlin!

I love the move of keeping all of what's common for custom type values while providing the extension point for what will vary. 👍🏼 

I need to reflect on the co-existence of the `represent` feature alongside the `StarlarkValueToGoValueConversion`. I _think_ it's okay, but I want to sleep on it.

Other than that, I'm really liking where this is landing.

**Comment by hanlins** [comment]
> This is fantastic, Hanlin!
> 
> I love the move of keeping all of what's common for custom type values while providing the extension point for what will vary. 👍🏼
> 
> I need to reflect on the co-existence of the `represent` feature alongside the `StarlarkValueToGoValueConversion`. I _think_ it's okay, but I want to sleep on it.
> 
> Other than that, I'm really liking where this is landing.

Hi @jtigger, thanks for the feedbacks! I think the tricky part for `StarlarkValueToGoValueConversion` is, as for now, it doesn't provide a way for people to emit errors. So is it ok if I refactor `StarlarkValueToGoValueConversion` the interface to allow it return error? If so, then I can make it return error such as `"urlValue: cannot coerce to a string; use .string()"`.

File: pkg/yamltemplate/filetests/ytt-library/url.yml
Comment (hanlins) [reply pkg/yamltemplate/filetests/ytt-library/url.yml]: There are no negative cases here. I can move the test cases to a separate file but I don't know how to add negative test cases.

Comment (pivotaljohn) [reply pkg/yamltemplate/filetests/ytt-library/url.yml]: Thanks for making that move, Hanlin.

The only negative case I can see is the representation error message. All the other cases are covered by a combination of tests I expect around `url.URL` and other ways we've confirmed that errors occurring through this code path make their way to the user.

FWIW, there _is_ a way to specify an error message in this parameterized test setup. The check is made here:
https://github.com/vmware-tanzu/carvel-ytt/blob/a1fa9e9bf578879c932ef360457c411f43ec2535/pkg/yamltemplate/template_test.go#L75-L82

And here's an example of it in use:
https://github.com/vmware-tanzu/carvel-ytt/blob/e850b3716926fae85418c652e50f4dcc42d0ca0d/pkg/yamltemplate/filetests/datatypes-err.yml#L9-L14

That said, I'd accept and merge this PR _without_ a specific test for this case.

**Review by pivotaljohn — APPROVED** [review]
LGTM. Thank you for all your work, here, Hanlin. We're not just adding a new feature, but setup the codebase to accept other custom types. 👍🏼

File: pkg/template/core/starlark_value.go
Comment (pivotaljohn) [inline pkg/template/core/starlark_value.go]: Regardless of how you decide that question "`represent` vs. `StarlarkValueToGoValueConversion` vs. ....?" (summarized above), seems like we can provide a significantly better experience to the end user if we return an `error` rather than panic out of `StarlarkValue.asInterface()`.

We'd love to preserve the panic for situations where we truly need to abort rather than as an error handling technique.

`AsGoValue()` already returns an `error`, so I think most of the plumbing is in place from there.

Doing so means that the internal Go stack trace will just go away, leaving the user with a short (and easier to read) error message.

File: pkg/yamltemplate/filetests/ytt-library/url-value.yml
Comment (pivotaljohn) [inline pkg/yamltemplate/filetests/ytt-library/url-value.yml]:
```suggestion
  test1: #@ testURL1.string()
  test2: #@ testURL1.user.string()
  test3: #@ testURL1.user.name
  test4: #@ testURL1.user.password
  test5: #@ testURL2.string()
  test6: #@ testURL2.user.string()
  test7: #@ testURL3.string()
  test8: #@ testURL3.user.string()
  test9: #@ testURL3.user.name
  test10: #@ testURL3.user.password
```

I created an inconsistency, here. The rationale that we came to for wanting to require that `urlValue` to no assume an encoding... well, not sure why it didn't occur to me that the same is true for `urlValue.user` 😬 .

File: pkg/yttlibrary/url.go
Comment (pivotaljohn) [inline pkg/yttlibrary/url.go]: _(optional)_ in this project, we're aiming to being careful in our API design _and_ defaulting to exporting primary types.  Given that the `urlValue` is effectively part of the `ytt` Starlark programming context, it's generally useful to export the same type in the Go code.

Suggestion: export `urlValue`

File: pkg/yttlibrary/url.go
Comment (pivotaljohn) [inline pkg/yttlibrary/url.go]: _(optional)_

is `urlUser` _actually_ a whole `urlValue`?
Would it make sense for it to hold a `*url.UserInfo`?

**Review by pivotaljohn — CHANGES_REQUESTED** [review]
## `represent` vs. `StarlarkValueToGoValueConversion` vs. ....?

Talked about this with @cppforlife and we arrived at what I thought was a cool compromise:

Given that the only error we'd be "throwing" is that the type is not convertible, what if we introduced an interface that specifically indicates the type is not convertible?

Something like:
```go
type UnconvertableStarlarkValue interface {
	ConversionHint() string
}
```

So, instead of muddying up `StarlarkValueToGoValueConversion` with an `error` (and all the error checks), a type would instead implement the `UnconvertableStarlarkValue` interface. `StarlarkValue.asInterface()` detects this interface and returns an `error` with part of the error message coming from `ConversionHint()`.

Personally, I'm digging how this makes the code more explicit about what's going on: that there are some types that are just not convertible... and that _those_ types ought to give contextualized hints/help in _how_ they might be converted.

Thoughts?

...

A few more comments in-line, below.

Note: a few thing I've marked as "_(optional)_". I'm sharing these as observations; I'm gonna leave it up to you whether you agree and want to include that change. _It's okay if you don't._

File: pkg/yttlibrary/url.go
Comment (hanlins) [reply pkg/yttlibrary/url.go]: I see. So apart of `UrlValue`, should we also export `urlUser` which is also a starlark type?

File: pkg/yttlibrary/url.go
Comment (hanlins) [reply pkg/yttlibrary/url.go]: Sure, fixed.

**Comment by hanlins** [comment]
> ## `represent` vs. `StarlarkValueToGoValueConversion` vs. ....?
> 
> Talked about this with @cppforlife and we arrived at what I thought was a cool compromise:
> 
> Given that the only error we'd be "throwing" is that the type is not convertible, what if we introduced an interface that specifically indicates the type is not convertible?
> 
> Something like:
> 
> ```go
> type UnconvertableStarlarkValue interface {
> 	ConversionHint() string
> }
> ```
> 
> So, instead of muddying up `StarlarkValueToGoValueConversion` with an `error` (and all the error checks), a type would instead implement the `UnconvertableStarlarkValue` interface. `StarlarkValue.asInterface()` detects this interface and returns an `error` with part of the error message coming from `ConversionHint()`.
> 
> Personally, I'm digging how this makes the code more explicit about what's going on: that there are some types that are just not convertible... and that _those_ types ought to give contextualized hints/help in _how_ they might be converted.
> 
> Thoughts?
> 
> ...
> 
> A few more comments in-line, below.
> 
> Note: a few thing I've marked as "_(optional)_". I'm sharing these as observations; I'm gonna leave it up to you whether you agree and want to include that change. _It's okay if you don't._

Thanks for the suggestions. I think it would make sense to have a separate changeset to cover this optional refactoring, as it's something invisible for consumers.

File: pkg/yamltemplate/filetests/ytt-library/url-value.yml
Comment (hanlins) [reply pkg/yamltemplate/filetests/ytt-library/url-value.yml]: Small correction here, in `test6: #@ testURL2.user.string()`, we actually should just keep it `test6: #@ testURL2.user` as `testURL2` doesn't contain user information, so its `user` member will be a `None` object.

**Review by pivotaljohn — APPROVED** [review]
This has been a journey in introducing a new kind of design element to `ytt`: a domain-specific Starlark type. You've been a terrific leading partner in this journey, @hanlins. Thank you for your efforts and commitment to getting us to this point. 🙏🏼

**Pull request merged**
